clippy_lints/src/uninit_vec.rs

```diff
@@ -0,0 +1,183 @@
+    /// undefined behavior with most safe operations.
+    /// Notably, using uninitialized `Vec<u8>` with generic `Read` is unsound.
+    ///
+    /// ### Example
+    /// ```rust,ignore
+    /// let mut vec: Vec<u8> = Vec::with_capacity(1000);
+    /// unsafe { vec.set_len(1000); }
+    /// reader.read(&mut vec); // undefined behavior!
+    /// ```
+    /// Use an initialized buffer:
+    /// ```rust,ignore
+    /// let mut vec: Vec<u8> = vec![0; 1000];
+    /// reader.read(&mut vec);
+    /// ```
+    /// Or, wrap the content in `MaybeUninit`:
+    /// ```rust,ignore
+    /// let mut vec: Vec<MaybeUninit<T>> = Vec::with_capacity(1000);
+    /// unsafe { vec.set_len(1000); }
+    /// ```
+    pub UNINIT_VEC,
+    correctness,
+    "Vec with uninitialized data"
+}
+
+declare_lint_pass!(UninitVec => [UNINIT_VEC]);
+
+impl<'tcx> LateLintPass<'tcx> for UninitVec {
+    fn check_block(&mut self, cx: &LateContext<'tcx>, block: &'tcx Block<'_>) {
+        for w in block.stmts.windows(2) {
+            if let StmtKind::Expr(expr) | StmtKind::Semi(expr) = w[1].kind {
+                handle_uninit_vec_pair(cx, &w[0], expr);
+            }
+        }
+
+        if let (Some(stmt), Some(expr)) = (block.stmts.last(), block.expr) {
+            handle_uninit_vec_pair(cx, stmt, expr);
+        }
+    }
+}
+
+fn handle_uninit_vec_pair(
+    cx: &LateContext<'tcx>,
+    maybe_with_capacity_or_reserve: &'tcx Stmt<'tcx>,
+    maybe_set_len: &'tcx Expr<'tcx>,
+) {
+    if_chain! {
+        if let Some(vec) = extract_with_capacity_or_reserve_target(cx, maybe_with_capacity_or_reserve);
+        if let Some((set_len_self, call_span)) = extract_set_len_self(cx, maybe_set_len);
+        if vec.eq_expr(cx, set_len_self);
+        if let ty::Ref(_, vec_ty, _) = cx.typeck_results().expr_ty_adjusted(set_len_self).kind();
+        if let ty::Adt(_, substs) = vec_ty.kind();
+        // Check T of Vec<T>
+        if !is_uninit_value_valid_for_ty(cx, substs.type_at(0));
+        then {
+            // FIXME: false positive #7698
+            #[allow(clippy::collapsible_span_lint_calls)]
+            span_lint_and_then(
+                cx,
+                UNINIT_VEC,
+                vec![call_span, maybe_with_capacity_or_reserve.span],
+                "calling `set_len()` immediately after reserving a buffer creates uninitialized values",
+                |diag| {
+                    diag.help("initialize the buffer or wrap the content in `MaybeUninit`");
+                },
+            );
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+enum LocalOrExpr<'tcx> {
+    Local(HirId),
+    Expr(&'tcx Expr<'tcx>),
+}
+
+impl<'tcx> LocalOrExpr<'tcx> {
+    fn eq_expr(self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) -> bool {
+        match self {
+            LocalOrExpr::Local(hir_id) => path_to_local_id(expr, hir_id),
+            LocalOrExpr::Expr(self_expr) => SpanlessEq::new(cx).eq_expr(self_expr, expr),
```

Comment:
I do think that it is more preferrable to check method calls in this version. Otherwise, it will cause false negatives when `vec` field is accessed through a method. It is rarer than a direct field access, but I think such use case exist. I mainly referred to [map_entry lint](https://rust-lang.github.io/rust-clippy/master/index.html#map_entry) when implementing this part, which uses `SpanlessEq` logic to detect `contains_key()` and `insert()` chain.

Maybe I don't have a full understanding of `SpanlessEq`. Could you provide an example of false positives? Is it like:
```rust
self.some_vec_returning_function().reserve(len);
self.some_vec_returning_function().set_len(len);
```
and the result of `some_vec_returning_function()` could be different in two calls? If so, I think the amount of false positive it will cause is negligible and it is better to check method calls by default. Or, is there some other false positives that I might not be aware of?

---

Thanks for the pointer in `Visitor` based implementation. I'll start working on them after handling PR comments for the `Send` lint.

Reply:
> and the result of some_vec_returning_function() could be different in two calls? If so, I think the amount of false positive it will cause is negligible and it is better to check method calls by default. Or, is there some other false positives that I might not be aware of?

No, that was the only FP that I was thinking of, that's also why it should be negligible. Writing a function like that would be bad practice anyway. So, I agree that using `SpanlessEq` is a good solution for this case :upside_down_face:.

Thank you for the detailed answers. You can resolve this :upside_down_face:.

Reply:
Thanks!